### PR TITLE
Add stat64 for arm and unify the check of fstat

### PR DESCRIPTION
**Pull request by bet4it**
### Which kind of PR do you create?

- [x] This PR only contains minor fixes.
- [ ] This PR contains major feature update.
- [ ] This PR introduces a new function/api for Qiling Framework.

### Coding convention?

- [x] The new code conforms to Qiling Framework naming convention.
- [x] The imports are arranged properly.
- [x] Essential comments are added.
- [x] The reference of the new code is pointed out.

### Extra tests?

- [ ] No extra tests are needed for this PR.
- [ ] I have added enough tests for this PR.
- [x] Tests will be added after some discussion and review.

### Changelog?

- [ ] This PR doesn't need to update Changelog.
- [x] Changelog will be updated after some proper review.
- [ ] Changelog has been updated in my PR.

### Target branch?

- [x] The target branch is dev branch.

### One last thing

- [x] I have read the [contribution guide](https://docs.qiling.io/en/latest/contribution/)

-----

The structure of stat64 for arm is not same with x86:
 https://github.com/qilingframework/qiling/blob/88a76ede81190409064fb1c9eb83811db672f2aa/qiling/os/posix/syscall/stat.py#L200-L204
https://github.com/qilingframework/qiling/blob/88a76ede81190409064fb1c9eb83811db672f2aa/qiling/os/posix/syscall/stat.py#L221-L225

Refer: 6b367403611db1261878480dc74a5386c1891fee and #346

### CODE REVIEW

**Comment by wtdcode** [comment]
Any reference? According to kernel source, they should be the same.

**Comment by bet4it** [comment]
Maybe this? https://elixir.bootlin.com/linux/latest/source/include/uapi/asm-generic/stat.h#L49

At least, I encounter errors when emulation and this fix works for me. And that's what 6b367403611db1261878480dc74a5386c1891fee want to solve.

**Comment by wtdcode** [comment]
Ah, I forgot stat64 before. You are right, but please add enough reference in comments like my previous code.

Thanks!

**Comment by bet4it** [comment]
You didn't forget stat64 before:
https://github.com/qilingframework/qiling/blob/d475262f72f933a395f190022e59aac5ee33bb47/qiling/os/posix/syscall/stat.py#L464
https://github.com/qilingframework/qiling/blob/d475262f72f933a395f190022e59aac5ee33bb47/qiling/os/posix/syscall/stat.py#L323

The thing we need to make out now is whether it uses [asm/stat.h](https://elixir.bootlin.com/linux/latest/source/arch/arm/include/uapi/asm/stat.h#L57) or [asm-generic/stat.h](https://elixir.bootlin.com/linux/latest/source/include/uapi/asm-generic/stat.h#L49) on arm.

**Comment by wtdcode** [comment]
I mean I forgot stat64 on arm64 several hours ago when I fixed stat on arm64. I will take a look tomorrow.

________________________________

**Comment by bet4it** [comment]
It seems that arm64 doesn't have stat64, stat64 only exists on arm32.

**Comment by wtdcode** [comment]
Yes, stat64 should only exist on 32bit systems as an extension to stat.

________________________________

**Comment by wtdcode** [comment]
> Maybe this? https://elixir.bootlin.com/linux/latest/source/include/uapi/asm-generic/stat.h#L49
> 
> At least, I encounter errors when emulation and this fix works for me. And that's what [6b36740](https://github.com/qilingframework/qiling/commit/6b367403611db1261878480dc74a5386c1891fee) want to solve.

What's your testcase? That reference seems an old `stat64` struct.

**Comment by bet4it** [comment]
Android arm32 program

File: qiling/os/posix/syscall/stat.py [hunk 1]
```diff
@@ -645,8 +645,32 @@ class LinuxARMEBStat(ctypes.BigEndianStructure):
...
+        ("__pad0", ctypes.c_uint8 * 4),
+        ("__st_ino", ctypes.c_uint32),
+        ("st_mode", ctypes.c_uint32),
+        ("st_nlink", ctypes.c_uint32),
+        ("st_uid", ctypes.c_uint32),
+        ("st_gid", ctypes.c_uint32),
+        ("st_rdev", ctypes.c_uint64),
+        ("__pad3", ctypes.c_uint8 * 8),
+        ("st_size", ctypes.c_int64),
+        ("st_blksize", ctypes.c_uint64),
```
Comment (wtdcode) [inline qiling/os/posix/syscall/stat.py:660]: why not `uint32` for `st_blksize`?

File: qiling/os/posix/syscall/stat.py [hunk 2]
```diff
@@ -645,8 +645,32 @@ class LinuxARMEBStat(ctypes.BigEndianStructure):
...
+    _fields_ = [
+        ("st_dev", ctypes.c_uint64),
+        ("__pad0", ctypes.c_uint8 * 4),
+        ("__st_ino", ctypes.c_uint32),
+        ("st_mode", ctypes.c_uint32),
+        ("st_nlink", ctypes.c_uint32),
+        ("st_uid", ctypes.c_uint32),
+        ("st_gid", ctypes.c_uint32),
+        ("st_rdev", ctypes.c_uint64),
+        ("__pad3", ctypes.c_uint8 * 8),
```
Comment (wtdcode) [inline qiling/os/posix/syscall/stat.py:658]: why 8 bytes?

**Review by wtdcode — COMMENTED** [review]
I made some investigation but still I can't find enough evidence to use the `stat64` structure you provide in your commits.

I suggest:

1. Provide your test binary && libs.
2. or provide a more solid evidence about the `stat64` structure. e.g. from kernel source, android source etc.

**Comment by wtdcode** [comment]
Fixed and add test case in 5dec91705be7b6d1124072d47b951439b2f6116c.

Thanks.

**Pull request closed without merging**